### PR TITLE
Install python-apt package if necessary

### PR DESCRIPTION
**Pull request by wtanaka**
This solves an issue where under certain configurations, ansible would
emit this error message:

```
msg: Could not import python modules: apt, apt_pkg. Please install
python-apt package.
```

### CODE REVIEW

**Comment by ahelal** [comment]
Thank you,

I think this might be problematic. It will break on non ubuntu system. If you're using an old Ansible can't you use `raw: "apt-get install python-apt -y"` in your play ?

**Comment by wtanaka** [comment]
On a non Ubuntu system, my intention is for the `-x` check to short circuit the execution of apt-get, and thus I expect this patch to maintain the previous behavior/error message of this code.

**Comment by ahelal** [comment]
That's cool, but I still see a case that this might break. If apt-get update was never run on the target, Then the **_apt-get install**_ will fail. So If you do decide todo `apt-get update` that would be costy and not elegant.
So my suggestion would be 
A) ignore the issue and delegate to your playbook or role. I usually run an apt-role in the start that takes care of apt that case could be easily added.
B) Check the host ansible version than 1- python-apt 2- update 3- install 
C) Upgrade Ansible :) 
D) If you have another idea I am open

**Comment by wtanaka** [comment]
that ^^^ was just a rebase with no other changes, I'll take a crack at option B

**Comment by ahelal** [comment]
Thank you.

I am currently on vacation will have a look at it late this week.

Regards
Adham

File: lib/kitchen/provisioner/ansible_push.rb [hunk 1]
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
       # For tests disable if not needed
       default_config :chef_bootstrap_url, "https://www.getchef.com/chef/install.sh"
 
+      attr_reader :conf
```
Comment (mcamou) [inline lib/kitchen/provisioner/ansible_push.rb]: Why have an `attr_reader :conf` if you already have a method called `conf`?

File: lib/kitchen/provisioner/ansible_push.rb [hunk 2]
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
...
+
+      # Validates the config and returns it.  Has side-effect of
+      # possibly setting @extra_vars which doesn't seem to be used
+      def conf
+        return @validated_config if defined? @validated_config
+
+        if !config[:playbook]
```
Comment (mcamou) [inline lib/kitchen/provisioner/ansible_push.rb:48]: could we change this to `config[:playbook].nil?` ?

File: lib/kitchen/provisioner/ansible_push.rb [hunk 3]
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
...
+          raise 'No playbook defined. Please specify one in .kitchen.yml'
+        end
+
+        if !File.exist?(config[:playbook])
+          raise "playbook '%s' could not be found. Please check path" % config[:playbook]
+        end
+
+        if config[:vault_password_file] and !File.exist?(config[:vault_password_file])
```
Comment (mcamou) [inline lib/kitchen/provisioner/ansible_push.rb:56]: Use `.nil?` and `unless` in a postfix position

File: lib/kitchen/provisioner/ansible_push.rb [hunk 4]
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
...
+          raise "Vault password '%s' could not be found. Please check path" % config[:vault_password_file]
+        end
+
+        # Validate that extra_vars is either a hash, or a path to an existing file
+        if config[:extra_vars]
```
Comment (mcamou) [inline lib/kitchen/provisioner/ansible_push.rb:61]: `.nil?`

File: lib/kitchen/provisioner/ansible_push.rb [hunk 5]
```diff
@@ -39,10 +40,108 @@ class AnsiblePush < Base
...
+            # Accept the usage of '@' (e.g. '@vars.yml' and 'vars.yml' are both supported)
+            match_data = /^@?(.+)$/.match(config[:extra_vars])
+            extra_vars_path = match_data[1].to_s
+            expanded_path = Pathname.new(extra_vars_path).expand_path(Dir.pwd)
+            extra_vars_is_valid = expanded_path.exist?
+            if extra_vars_is_valid
+              @extra_vars = '@' + extra_vars_path
+            end
+          end
+          if !extra_vars_is_valid
```
Comment (mcamou) [inline lib/kitchen/provisioner/ansible_push.rb:73]: use `unless` and put in postfix position

File: lib/kitchen/provisioner/ansible_push.rb [hunk 6]
```diff
@@ -50,12 +149,19 @@ def prepare_command
...
-        omnibus_download_dir = config[:omnibus_cachier] ? "/tmp/vagrant-cache/omnibus_chef" : "/tmp"
-        chef_url = config[:chef_bootstrap_url]
+        stdin, stdout, stderr = Open3.popen3(command_env(), command() + " --version") 
+        version_output = stdout.read()
+        version_string = version_output.split()[1]
+
+        omnibus_download_dir = conf[:omnibus_cachier] ? "/tmp/vagrant-cache/omnibus_chef" : "/tmp"
+        chef_url = conf[:chef_bootstrap_url]
+
         if chef_url
```
Comment (mcamou) [inline lib/kitchen/provisioner/ansible_push.rb:155]: `unless chef_url.nil?`

File: lib/kitchen/provisioner/ansible_push.rb (hunk 1 above)
Comment (wtanaka) [inline lib/kitchen/provisioner/ansible_push.rb]: Syntactic sugar, I can remove it if you prefer; code would look like 

conf()[:some_key]

?

File: lib/kitchen/provisioner/ansible_push.rb [hunk 7]
```diff
@@ -123,111 +251,41 @@ def exec_ansible_command(env, command, desc)
...
-          "PYTHONUNBUFFERED" => "1", # Ensure Ansible output isn't buffered
-          "ANSIBLE_FORCE_COLOR" => "true",
-          "ANSIBLE_HOST_KEY_CHECKING" => "#{config[:host_key_checking]}",
-        }
-        @command_env["ANSIBLE_CONFIG"]=config[:ansible_config] if config[:ansible_config]
-        info("Ansible push compile conig done")
-      end
-
-      def validate_config()
-        if !config[:playbook]
```
Comment (wtanaka) [inline lib/kitchen/provisioner/ansible_push.rb:181]: @mcamou  all this code already existed, I just moved it around and extracted functions, it doesn't seem like making further changes of coding style would be relevant to this specific pull request?

File: lib/kitchen/provisioner/ansible_push.rb (hunk 6 above)
Comment (wtanaka) [inline lib/kitchen/provisioner/ansible_push.rb:155]: I didn't touch this line either way in this pull request

File: lib/kitchen/provisioner/ansible_push.rb (hunk 7 above)
Comment (ahelal) [inline lib/kitchen/provisioner/ansible_push.rb:181]: @mcamou @wtanaka I would take care of that part.

File: lib/kitchen/provisioner/ansible_push.rb (hunk 1 above)
Comment (ahelal) [inline lib/kitchen/provisioner/ansible_push.rb]: but isn't it true that in Ruby when calling a no-argument method you don't have to specify parentheses? For example:

```
irb(main):004:0> def x
irb(main):005:1> { :a => 1, :b => 2 }
irb(main):006:1> end
=> :x
irb(main):007:0> x[:a]
=> 1
```

To my understanding  `attr_reader conf``and``attr_reader :machine_name` is redundant

Comment (wtanaka) [inline lib/kitchen/provisioner/ansible_push.rb]: @ahelal Removed and updated pull request

**Pull request merged**